### DISCUSSION TITLE
Research Comparison Report: ONNX vs Python ZMQ Model Service. ONNX way faster!!

### DISCUSSION THREAD

**Discussion by Lqz13Th** [Lqz13Th/extrema_infra#15] [Ideas]
Research Comparison Report: ONNX vs Python ZMQ Model Service

Document version: v1.0
Benchmark suite: onnx_zmq_benchmark_suite v1
Runner package version: 0.1.0
Generated time (Asia/Shanghai): 2026-04-06T16:05:17+08:00
Generated time (UTC): 2026-04-06T08:05:17Z

1. Objective
- This report compares two inference paths on the same machine:
  1) direct in-process ONNX execution from Rust using tract-onnx
  2) Rust -> localhost ZMQ -> Python model server
- The benchmark scope contains three requested dimensions:
  1) minimal-path comparison
  2) feature-scale comparison (10 / 50 / 500 features)
  3) model-complexity comparison (small / medium / large)

2. Publication Summary
- Across all 12 matched cases, mean ZMQ latency is 35.42x mean ONNX latency.
- The smallest ZMQ/ONNX ratio is lgbm_f50_large at 2.00x.
- The largest ZMQ/ONNX ratio is mlp_f500_medium at 142.25x.
- The minimal-compare section averages 28.22x.
- In the feature-scale section, the MLP average ratio is 81.34x and the LGBM average ratio is 7.09x.
- In the model-complexity section, the MLP average ratio is 51.40x and the LGBM average ratio is 7.39x.

3. Version and Environment
- Rust: rustc 1.94.0 (4a4ef493e 2026-03-02)
- Cargo: cargo 1.94.0 (85eff7c80 2026-01-15)
- Python model environment: Python 3.12.7 (model_examples/.venv)
- Python transport environment: Python 3.12.7 (base interpreter)
- Timezone: Asia/Shanghai
- Build profile: release
- Measured benchmark duration: 20.13 seconds
- ZMQ port: 5552

4. Model Libraries, Runtime Stack, and Dependency Versions
- Rust-side dependencies:
  - tract-onnx 0.22.1
  - zeromq 0.5.0
  - tokio 1.51.0
  - rmp-serde 1.3.1
  - serde 1.0.228
  - serde_json 1.0.149
- Python model-side dependencies:
  - torch 2.11.0
  - lightgbm 4.6.0
  - hummingbird-ml 0.4.12
  - numpy 2.4.4
  - onnx 1.16.1
  - onnxconverter_common 1.16.0
  - onnxmltools 1.16.0
  - scikit-learn 1.8.0
  - skl2onnx 1.20.0
- Python transport / serialization dependencies:
  - pyzmq 25.1.2
  - libzmq 4.3.5
  - msgpack 1.0.3
- Implementation notes:
  - The ONNX runtime used here is tract-onnx 0.22.1, not onnxruntime.
  - The LGBM ONNX export path is Hummingbird -> torch -> ONNX.
  - ONNX export uses torch.onnx.export with opset_version=17 and dynamo=False (legacy TorchScript exporter).
  - The ZMQ service script falls back to the base Python environment for pyzmq/msgpack if those packages are not present in the model venv.

5. Commands Executed
  cargo check --manifest-path rust_benchmark_suite/Cargo.toml --offline
  model_examples\.venv\Scripts\python.exe benchmarks\build_suite_artifacts.py
  cargo run --release --manifest-path rust_benchmark_suite/Cargo.toml --offline -- --warmup-iters 40 --measure-iters 300 --repeats 8 --report-label detailed_suite

6. Benchmark Methodology and Measurement Details
- warmup_iters=40, measure_iters=300, repeats=8.
- Effective measured samples per case + runner = 2400.
- ONNX path: Rust preloads the ONNX graph, optimizes it, converts it into a runnable model, and executes a single input row in-process.
- ZMQ path: Rust prebuilds the msgpack request and sends it through localhost REQ/REP to a preloaded Python model server.
- Input batch size is one sample; tabular input shape is [1, feature_count]; output shape is [1, 1].
- Python MLP inference uses torch.inference_mode(); LightGBM inference uses native Booster.predict().
- Torch training and serving are forced to single-thread mode; LightGBM training uses num_threads=1.
- Synthetic dataset sizes: TRAIN_ROWS=4096, VALID_ROWS=256.
- The synthetic target mixes a linear signal, tanh nonlinearity, low-order interactions, a sine term, and Gaussian noise.
- MLP training parameters: Adam, lr=0.01, epochs=120.
- LGBM fixed hyperparameters include learning_rate=0.08, min_data_in_leaf=4, min_data_in_bin=4, subsample=0.95, and colsample_bytree=0.95.
- Output validation tolerances: MLP abs tol=1e-3, LGBM abs tol=1e-2.
- A 25 ms pause is inserted between repeats to reduce back-to-back run jitter.

7. Benchmark Matrix (all 12 cases)
- mlp_f10_small: title=MLP small / 10 features, tags=minimal_compare, description=PyTorch MLP regressor with hidden_layers=[16] and 10 tabular input features., input_shape=[1, 10], output_shape=[1, 1], hidden_layers=[16], params=193, onnx=1.26 KB, python_artifact=3.27 KB
- lgbm_f10_small: title=LGBM small / 10 features, tags=minimal_compare, description=LightGBM regressor with n_estimators=16, max_depth=4, num_leaves=15, exported through Hummingbird -> torch -> ONNX., input_shape=[1, 10], output_shape=[1, 1], trees=16, depth=4, leaves=15, onnx=10.43 KB, python_artifact=25.94 KB
- mlp_f10_medium: title=MLP medium / 10 features, tags=feature_scale, description=PyTorch MLP regressor with hidden_layers=[64, 32] and 10 tabular input features., input_shape=[1, 10], output_shape=[1, 1], hidden_layers=[64, 32], params=2817, onnx=11.79 KB, python_artifact=14.10 KB
- mlp_f50_medium: title=MLP medium / 50 features, tags=feature_scale,model_complexity, description=PyTorch MLP regressor with hidden_layers=[64, 32] and 50 tabular input features., input_shape=[1, 50], output_shape=[1, 1], hidden_layers=[64, 32], params=5377, onnx=21.79 KB, python_artifact=24.10 KB
- mlp_f500_medium: title=MLP medium / 500 features, tags=feature_scale, description=PyTorch MLP regressor with hidden_layers=[64, 32] and 500 tabular input features., input_shape=[1, 500], output_shape=[1, 1], hidden_layers=[64, 32], params=34177, onnx=134.29 KB, python_artifact=136.61 KB
- lgbm_f10_medium: title=LGBM medium / 10 features, tags=feature_scale, description=LightGBM regressor with n_estimators=48, max_depth=6, num_leaves=31, exported through Hummingbird -> torch -> ONNX., input_shape=[1, 10], output_shape=[1, 1], trees=48, depth=6, leaves=31, onnx=57.49 KB, python_artifact=136.60 KB
- lgbm_f50_medium: title=LGBM medium / 50 features, tags=feature_scale,model_complexity, description=LightGBM regressor with n_estimators=48, max_depth=6, num_leaves=31, exported through Hummingbird -> torch -> ONNX., input_shape=[1, 50], output_shape=[1, 1], trees=48, depth=6, leaves=31, onnx=57.49 KB, python_artifact=138.92 KB
- lgbm_f500_medium: title=LGBM medium / 500 features, tags=feature_scale, description=LightGBM regressor with n_estimators=48, max_depth=6, num_leaves=31, exported through Hummingbird -> torch -> ONNX., input_shape=[1, 500], output_shape=[1, 1], trees=48, depth=6, leaves=31, onnx=57.49 KB, python_artifact=166.46 KB
- mlp_f50_small: title=MLP small / 50 features, tags=model_complexity, description=PyTorch MLP regressor with hidden_layers=[16] and 50 tabular input features., input_shape=[1, 50], output_shape=[1, 1], hidden_layers=[16], params=833, onnx=3.76 KB, python_artifact=5.77 KB
- lgbm_f50_small: title=LGBM small / 50 features, tags=model_complexity, description=LightGBM regressor with n_estimators=16, max_depth=4, num_leaves=15, exported through Hummingbird -> torch -> ONNX., input_shape=[1, 50], output_shape=[1, 1], trees=16, depth=4, leaves=15, onnx=10.43 KB, python_artifact=28.16 KB
- mlp_f50_large: title=MLP large / 50 features, tags=model_complexity, description=PyTorch MLP regressor with hidden_layers=[256, 128, 64] and 50 tabular input features., input_shape=[1, 50], output_shape=[1, 1], hidden_layers=[256, 128, 64], params=54273, onnx=213.07 KB, python_artifact=215.66 KB
- lgbm_f50_large: title=LGBM large / 50 features, tags=model_complexity, description=LightGBM regressor with n_estimators=96, max_depth=8, num_leaves=63, exported through Hummingbird -> torch -> ONNX., input_shape=[1, 50], output_shape=[1, 1], trees=96, depth=8, leaves=63, onnx=396.51 KB, python_artifact=543.21 KB

8. Section Results
- Minimal Compare: avg_mean_ratio=28.22x, range=13.48x~42.95x, avg_p99_ratio=46.64x
- Feature Scale / MLP: avg_mean_ratio=81.34x, range=33.07x~142.25x, avg_p99_ratio=155.11x
- Feature Scale / LGBM: avg_mean_ratio=7.09x, range=4.04x~12.87x, avg_p99_ratio=4.92x
- Model Complexity / MLP: avg_mean_ratio=51.40x, range=26.40x~68.70x, avg_p99_ratio=105.05x
- Model Complexity / LGBM: avg_mean_ratio=7.39x, range=2.00x~15.83x, avg_p99_ratio=6.59x

8.1 Minimal Compare
- mlp_f10_small: family=mlp, features=10, complexity=small, meta=(hidden_layers=[16], params=193), ONNX mean=4.47 us, ZMQ mean=191.78 us, ratio=42.95x, ONNX p99=5.70 us, ZMQ p99=446.00 us, ONNX qps=223964.17, ZMQ qps=5214.33, artifacts=(onnx=1.26 KB, python_artifact=3.27 KB)
- lgbm_f10_small: family=lgbm, features=10, complexity=small, meta=(trees=16, depth=4, leaves=15), ONNX mean=24.07 us, ZMQ mean=324.55 us, ratio=13.48x, ONNX p99=33.00 us, ZMQ p99=496.30 us, ONNX qps=41544.05, ZMQ qps=3081.20, artifacts=(onnx=10.43 KB, python_artifact=25.94 KB)

8.2 Feature Scale / MLP
- mlp_f10_medium: family=mlp, features=10, complexity=medium, meta=(hidden_layers=[64, 32], params=2817), ONNX mean=6.43 us, ZMQ mean=212.80 us, ratio=33.07x, ONNX p99=7.70 us, ZMQ p99=505.50 us, ONNX qps=155401.16, ZMQ qps=4699.18, artifacts=(onnx=11.79 KB, python_artifact=14.10 KB)
- mlp_f50_medium: family=mlp, features=50, complexity=medium, meta=(hidden_layers=[64, 32], params=5377), ONNX mean=4.23 us, ZMQ mean=290.92 us, ratio=68.70x, ONNX p99=4.40 us, ZMQ p99=665.50 us, ONNX qps=236132.16, ZMQ qps=3437.31, artifacts=(onnx=21.79 KB, python_artifact=24.10 KB)
- mlp_f500_medium: family=mlp, features=500, complexity=medium, meta=(hidden_layers=[64, 32], params=34177), ONNX mean=5.02 us, ZMQ mean=713.61 us, ratio=142.25x, ONNX p99=5.50 us, ZMQ p99=1366.40 us, ONNX qps=199343.83, ZMQ qps=1401.32, artifacts=(onnx=134.29 KB, python_artifact=136.61 KB)

8.3 Feature Scale / LGBM
- lgbm_f10_medium: family=lgbm, features=10, complexity=medium, meta=(trees=48, depth=6, leaves=31), ONNX mean=77.24 us, ZMQ mean=311.97 us, ratio=4.04x, ONNX p99=171.00 us, ZMQ p99=498.80 us, ONNX qps=12947.09, ZMQ qps=3205.40, artifacts=(onnx=57.49 KB, python_artifact=136.60 KB)
- lgbm_f50_medium: family=lgbm, features=50, complexity=medium, meta=(trees=48, depth=6, leaves=31), ONNX mean=85.53 us, ZMQ mean=372.10 us, ratio=4.35x, ONNX p99=147.00 us, ZMQ p99=509.20 us, ONNX qps=11691.53, ZMQ qps=2687.48, artifacts=(onnx=57.49 KB, python_artifact=138.92 KB)
- lgbm_f500_medium: family=lgbm, features=500, complexity=medium, meta=(trees=48, depth=6, leaves=31), ONNX mean=82.06 us, ZMQ mean=1055.94 us, ratio=12.87x, ONNX p99=152.50 us, ZMQ p99=1276.70 us, ONNX qps=12186.45, ZMQ qps=947.02, artifacts=(onnx=57.49 KB, python_artifact=166.46 KB)

8.4 Model Complexity / MLP @ 50 features
- mlp_f50_small: family=mlp, features=50, complexity=small, meta=(hidden_layers=[16], params=833), ONNX mean=4.19 us, ZMQ mean=247.64 us, ratio=59.11x, ONNX p99=5.10 us, ZMQ p99=607.40 us, ONNX qps=238699.09, ZMQ qps=4038.13, artifacts=(onnx=3.76 KB, python_artifact=5.77 KB)
- mlp_f50_medium: family=mlp, features=50, complexity=medium, meta=(hidden_layers=[64, 32], params=5377), ONNX mean=4.23 us, ZMQ mean=290.92 us, ratio=68.70x, ONNX p99=4.40 us, ZMQ p99=665.50 us, ONNX qps=236132.16, ZMQ qps=3437.31, artifacts=(onnx=21.79 KB, python_artifact=24.10 KB)
- mlp_f50_large: family=mlp, features=50, complexity=large, meta=(hidden_layers=[256, 128, 64], params=54273), ONNX mean=10.09 us, ZMQ mean=266.32 us, ratio=26.40x, ONNX p99=13.60 us, ZMQ p99=609.40 us, ONNX qps=99137.91, ZMQ qps=3754.86, artifacts=(onnx=213.07 KB, python_artifact=215.66 KB)

8.5 Model Complexity / LGBM @ 50 features
- lgbm_f50_small: family=lgbm, features=50, complexity=small, meta=(trees=16, depth=4, leaves=15), ONNX mean=23.83 us, ZMQ mean=377.12 us, ratio=15.83x, ONNX p99=35.70 us, ZMQ p99=521.80 us, ONNX qps=41969.12, ZMQ qps=2651.66, artifacts=(onnx=10.43 KB, python_artifact=28.16 KB)
- lgbm_f50_medium: family=lgbm, features=50, complexity=medium, meta=(trees=48, depth=6, leaves=31), ONNX mean=85.53 us, ZMQ mean=372.10 us, ratio=4.35x, ONNX p99=147.00 us, ZMQ p99=509.20 us, ONNX qps=11691.53, ZMQ qps=2687.48, artifacts=(onnx=57.49 KB, python_artifact=138.92 KB)
- lgbm_f50_large: family=lgbm, features=50, complexity=large, meta=(trees=96, depth=8, leaves=63), ONNX mean=186.65 us, ZMQ mean=374.14 us, ratio=2.00x, ONNX p99=327.70 us, ZMQ p99=554.60 us, ONNX qps=5357.71, ZMQ qps=2672.81, artifacts=(onnx=396.51 KB, python_artifact=543.21 KB)

9. Output Files
- report.md
- summary.json
- summary.csv
- pair_summary.json
- pair_summary.csv
- repeat_summary.json
- repeat_summary.csv
- section_summary.json
- section_summary.csv
- raw_results.json
- manifest_snapshot.json

10. Conclusions and Notes
- Under this single-machine, single-sample, single-thread setup, the extra ZMQ cost mainly comes from serialization, socket round-trip, and the Python service boundary.
- For lightweight MLP models, communication and boundary-crossing overhead dominate, so the ZMQ/ONNX ratio expands dramatically.
- For heavier LGBM models, compute dominates a larger share of total latency, so the ZMQ/ONNX ratio compresses into a lower range.
- 500-feature inputs amplify the ZMQ path cost substantially, especially in the MLP section.
- These results are most reliable for relative comparison on the same machine and setup. Absolute latency numbers may change on different hardware, threading, or transport stacks, but the trend is still informative.

11. Reproduction Suggestions
- Regenerate the model artifacts first, then rerun the benchmark in release mode.
- If the experiment needs to be extended further, the most natural next axes are batch size, concurrency, serialization format, and alternative IPC/HTTP transports.
